components/raftstore/src/store/util.rs

```diff
@@ -927,11 +936,36 @@ pub struct ReadState {
 }
 
 impl RegionReadProgressCore {
-    fn new(applied_index: u64, cap: usize) -> RegionReadProgressCore {
+    fn new(applied_index: u64, cap: usize, tag: String) -> RegionReadProgressCore {
         RegionReadProgressCore {
+            tag,
             applied_index,
             read_state: ReadState::default(),
             pending_items: VecDeque::with_capacity(cap),
+            last_merge_index: 0,
+        }
+    }
+
+    // Reset `safe_ts` to min(`source_safe_ts`, `safe_ts`)
+    fn merge_safe_ts(&mut self, source_safe_ts: u64, merge_index: u64) -> Option<u64> {
+        // Consume all pending items before `merge_index`
+        self.update_applied(merge_index);
+        // Update `last_merge_index`
+        self.last_merge_index = merge_index;
+        // Reset target region's `safe_ts`
+        let target_safe_ts = self.read_state.ts;
+        self.read_state.ts = cmp::min(source_safe_ts, target_safe_ts);
+        info!(
+            "reset safe_ts due to merge";
+            "tag" => &self.tag,
+            "source_safe_ts" => source_safe_ts,
+            "target_safe_ts" => target_safe_ts,
+            "safe_ts" => self.read_state.ts,
+        );
```

Comment:
In fact, I think the log may not be helpful if we didn't print the region-id here.

Reply:
The region id and peer id is contained in `self.tag`